src/main/java/org/datadog/jenkins/plugins/datadog/util/git/GitMetadataBuilderCallback.java

```diff
@@ -75,36 +87,39 @@ private String getDefaultBranch(Repository repository, String remoteName) throws
         return null;
     }
 
-    private static String getBranch(Repository repository) throws IOException {
+    private static Collection<String> getBranches(Repository repository) throws IOException {
+        List<String> matchingBranches = new ArrayList<>();
+
         String branch = repository.getBranch();
         if (!GitUtils.isValidCommitSha(branch)) {
-          return branch;
+            matchingBranches.add(branch);
+            return matchingBranches;
         }
 
         // A detached HEAD is checked out.
-        // Iterate over available refs to see if any of them points to the checked out commit.
+        // Iterate over available refs to see which point to the checked out commit (there may be multiple).
         for (Ref ref : repository.getRefDatabase().getRefs()) {
             String refName = ref.getName();
             if (Constants.HEAD.equals(refName) || GitUtils.isValidCommitSha(refName)) {
                 continue;
             }
             ObjectId refObjectId = ref.getObjectId();
             if (branch.equals(refObjectId.getName())) {
-                return refName;
+                matchingBranches.add(normalizeBranch(refName));
             }
         }
-        return null;
+        return matchingBranches;
```

Comment:
This method now is basically iterating over _**all**_ available refs in the repository. Is this expected? 
I foresee issues with this. The number of iterations can be huge depending on the repository.

Also, it seems the method `chooseBranch(...)` is always getting the first branch in the collection.

Am I missing something?

Reply:
It does iterate over refs in the repo, but that was possible before the changes too, depending on where in the list the branch that we're looking for was located.

The method returns the first branch if the collection does not contain the name of the branch that is set in the environment variables (which may or may not be the case, especially with pipelines that do multiple checkouts - multi-branch pipelines, pipelines that use shared libraries, etc).

Anyway, I rewrote the logic a bit to address the comments. Now branch resolution is done inside the callback, and examining the references can be disabled with an env var.